Allow moving tracks within queue using track menu

Chaphasilor opened the pull request:
Implements #1434

<!-- Just a reminder that text in these brackets is not visible in the rendered output.
You also do **not** need to use this template. You can remove and modify anything however you like!
Its just a suggestion :)
-->

## Changes
<!-- Please describe what this Pull request adds or changes. Possibly with images -->
- Added `PlaybackActionRow` if needed
- Implemented move logic via removing and re-adding
  - Simply reordering doesn't work since we can't change a `FinampQueueItem`'s `QueueItemQueueType` after it was added to the queue

## Todo before merging
<!-- Add custom todos if deemed necessary to give others an overview on how far this PR is progressing -->

<!-- Did you add UI text? 
No? Just check or remove the box
Yes? Make sure you replace hardcoded strings with translations -->
- [x] Translations

## Related Issues
<!-- List relevant issues to this PR using a format like this
- closes #xyz 
- fixes #abc
- follow up to #pr-name
-->
- closes #1434

Chaphasilor committed:
add `PlaybackActionRow` with move operations to `QueueListTile`'s track menu

Chaphasilor commented:
@Komodo5197 I assume no comments from you means this can be merged?

Komodo5197 commented on lib/menus/components/playbackActions/playback_actions.dart:
```diff
@@ -19,12 +19,65 @@ Map<String, Widget> getPlaybackActionPages({
+            ],
+          ),
+          // New Queue
+          AppLocalizations.of(context)!.playbackActionPageNewQueue: Row(
```
Is it really worth it to give this option its own page?  This is also inconsistent with the page layout of the same four items for non-queue items.

Komodo5197 commented on lib/menus/components/playbackActions/playback_actions.dart:
```diff
@@ -19,12 +19,65 @@ Map<String, Widget> getPlaybackActionPages({
   final BaseItemDtoType? itemType = item is BaseItemDto ? BaseItemDtoType.fromItem(item) : null;
 
-  return {
+  return queueItem != null
```
We seem to be using queueItem as a proxy for whether we have a track.  I'd prefer to directly check if the item is a track, and then show or hide the move page based off the queueItem, to make this more understandable and robust.

Komodo5197 commented on lib/menus/track_menu.dart:
```diff
@@ -476,13 +481,15 @@ class _TrackMenuState extends ConsumerState<TrackMenu> with TickerProviderStateM
         height: MenuItemInfoSliverHeader.defaultHeight,
         child: SliverToBoxAdapter(
-          child: Row(
+          child: widget.queueItem != null
```
I'd prefer to just use the PlaybackActionRow for non-queue items as well, for more consistency.  We'd need to hide the page headers for one-page rows, though, so you can leave them seperate if you'd prefer.  I would like the action page structure to be consistent though.

Chaphasilor replied:
good point, we can just hide the header instead of doing completely manually 👍

Komodo5197 commented on lib/models/finamp_models.dart:
```diff
@@ -2155,6 +2155,23 @@ class FinampQueueInfo {
     return null;
   }
 
+  FinampQueueItem? getQueueItemByOffset(int offset) {
```
Is this actually used?

Chaphasilor replied:
no, I just included it for completeness' sake, I think. Or I removed the code that used it. I wouldn't mind keeping it around though, might be useful

Komodo5197 commented on lib/models/finamp_models.dart:
```diff
@@ -2155,6 +2155,23 @@ class FinampQueueInfo {
+    return fullQueue[index];
+  }
+
+  int? getOffsetForQueueItem(FinampQueueItem item) {
```
It might make more sense to just directly add a removeQueueItem method to the queue service?

Komodo5197 commented on lib/l10n/app_en.arb:
```diff
@@ -1054,6 +1062,18 @@
   "@shuffleSomeToQueue": {
     "description": "Used for shuffling some tracks to the end of the regular queue, to play after all prior tracks from the queue have played"
   },
+  "movePlayNext": "Play next",
```
Could we really not come up with a name that isn't completely identical to the insert variant?  although I'm not coming up with anything that good either...

Chaphasilor replied:
Question is: do we need a different name at all? Semantically it's different, sure, but the item will actually play next, so...

Komodo5197 commented on lib/menus/components/playbackActions/playback_action_row.dart:
```diff
@@ -51,7 +53,8 @@ class _PlaybackActionRowState extends ConsumerState<PlaybackActionRow> {
       popContext: widget.popContext,
       compactLayout: widget.compactLayout,
       genreFilter: widget.genreFilter,
-      preferNextUp: ref.watch(finampSettingsProvider.preferNextUpPrepending),
```
You should create a separate last used page setting for queue tracks, because the page types don't really line up with the other action rows.

Komodo5197 commented:
I think I just never got around to reading through this thoroughly?  I've got a couple minor issues and nitpicks, but the big thing is that I think this needs to remember its last used page independently of the other playback rows.

Chaphasilor commented:
Alright, thanks! Will try to get this finished up tonight

Chaphasilor committed:
format

Chaphasilor committed:
remember pages separately, hide header for single page, simplify logic, refactor

Chaphasilor commented:
Okay, I think I addressed all the points now, sorry for the delay. I also tried to simplify the page->index and index->page lookups, and I believe the behavior is still the same/desired one.  
Right now the track menu uses 3 or 4 pages instead of the single page we had before, while this is more consistent with the other options, it potentially leads to more taps needed to perform an action, so I'm not yet sure if we should change that. I guess we could release it and see if people want the "old" behavior back?

Komodo5197 commented:
I was talking more about similarity between track menu in the queue, with the move option, and the track menu outside the queue without it.  I don't think trying to be consistent with non-tracks is worthwhile because tracks are to different, as they don't have the shuffle options.  The design I'd go with would have the move page, if applicable, and then a page with all four other track options.  I honestly think our playback rows are already too split up for albums and whatnot, with only two options per row, but I don't think we have a way use less groups and still have meaningful divisions.  Anyway, I don't have a catchy name for the track monopage, but I think we should keep it.  Also, I think there's something wrong with the stackSized calculation in the track menu, because I'm occasionally getting scrollbars.

Chaphasilor committed:
revert to consolidated playback action row page for tracks, fix stack height for track menu

Chaphasilor commented:
@Komodo5197 okay, yeah that makes sense, I agree. That has been reverted now, and I think I've also fixed the height issue for the track menu. I think by "getting scrollbars" you meant the bottom sheet not being tall enough to fix all options? If so, that should be fixed.

For the second page that has all the regular options for the track menu, I've decided to just call it "Play", because while I originally wanted to differentiate between "duplicate" (i.e., play next and add to Next Up) and "new queue" (i.e., play), that doesn't work for a single page, and is also not that important. Having "move" and "play" as options should be clear enough.

Chaphasilor committed:
review findings

Chaphasilor commented:
@Komodo5197 did you have a look at the new changes yet? Can we merge?

Komodo5197 commented:
I haven't bothered redoing any testing, but it all looks good from the code.  With the just_audio PR merged, make sure this doesn't accidentally disable any of the next up playback options that should be hopefully working on desktop now.

Chaphasilor committed:
Merge branch 'redesign' into redesign-move-to-next-up-within-queue, make sure Next Up and shuffle are enabled on desktop

Chaphasilor committed:
format files
